### PR TITLE
make type_name() optional

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/iresearch_s-pr/142/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/job/iresearch-release-s/1156/

File: core/utils/std.hpp [hunk 1]
```diff
@@ -28,6 +28,22 @@
...
+    using no_type = long;                                                  \
+    template<typename U> static yes_type test(decltype(&U::member_name));  \
+    template<typename U> static no_type  test(...);                        \
+   public:                                                                 \
+    static constexpr bool value = sizeof(test<T>(0)) == sizeof(yes_type);  \
+  };                                                                       \
+  template<typename T>                                                     \
+  inline constexpr auto has_member_##member##_v = has_member_##member<T>::value
+
+#define HAS_MEMBER(type, member) has_member_##member_v<type>
```
Comment (Dronplane) [inline core/utils/std.hpp:45]:
```suggestion
#define HAS_MEMBER(type, member) has_member_##member##_v<type>
```

File: core/utils/std.hpp [hunk 2]
```diff
@@ -28,6 +28,22 @@
...
+#define DEFINE_HAS_MEMBER(member)                                          \
+  template<typename T>                                                     \
+  class has_member_##member {                                              \
+   private:                                                                \
+    using yes_type = char;                                                 \
+    using no_type = long;                                                  \
+    template<typename U> static yes_type test(decltype(&U::member_name));  \
+    template<typename U> static no_type  test(...);                        \
+   public:                                                                 \
+    static constexpr bool value = sizeof(test<T>(0)) == sizeof(yes_type);  \
```
Comment (Dronplane) [inline core/utils/std.hpp:40]:
```suggestion
    static constexpr bool value = sizeof(test<typename std::remove_reference<typename std::remove_cv<T>::type>::type>(0)) == sizeof(yes_type);  \
```

File: core/utils/std.hpp [hunk 3]
```diff
@@ -28,6 +28,22 @@
 
 #include "shared.hpp"
 
+#define DEFINE_HAS_MEMBER(member)                                          \
+  template<typename T>                                                     \
+  class has_member_##member {                                              \
+   private:                                                                \
+    using yes_type = char;                                                 \
+    using no_type = long;                                                  \
+    template<typename U> static yes_type test(decltype(&U::member_name));  \
```
Comment (Dronplane) [inline core/utils/std.hpp:37]:
```suggestion
    template<typename U> static yes_type test(decltype(&U::member));  \
```

**Review by Dronplane — CHANGES_REQUESTED** [review]

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/iresearch_s-pr/144/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/iresearch_s-pr/145/

**Comment by gnusi** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/iresearch_s-pr/148/

**Pull request merged**